### DISCUSSION TITLE
how to transfer an allocated array on GPU to solver-appl

### DISCUSSION THREAD

**Discussion by ztdepztdep** [ginkgo-project/ginkgo#1131] [Q&A] [answered]
The d_b and d_x is two double arraies which have been allocated on GPU. I want to know how to transfer it to the solver-apply.
I tried to use the following code:
solver->apply(gko::lend(d_b), gko::lend(d_x));
but it doesn't wok.

**Comment by yhmtsai** [comment]
Are the type of `d_b` and `d_x` `double*`?
If they are, the solver needs to use the `gko::matrix::Dense<double>*` type.
You can create the Dense matrix by reading a mtx file.
ex. `gko::read<gko::matrix::Dense<>>(std::ifstream("filepath"), exec);`
or create the Dense first then fill in the data
```
auto x = gko::matrix::Dense<>::create(exec, gko::dim<2>(n, 1));
// copy the data to x->get_values()
```
or create the Dense on the preallocated array
`auto x = gko::matrix::Dense<>::create(exec, gko::dim<2>(n, 1), gko::array<double>::view(exec, n, d_x));`
note. array view to indicate the size and location for the plain pointer.

**Comment by yhmtsai** [comment, marked as answer]
there is another discussion transfer from issue, which is the same discussion as this.
I will lock this conversation, please go to another one https://github.com/ginkgo-project/ginkgo/discussions/1132
